### PR TITLE
Issue/236 modules v2 plugin loading

### LINKED ISSUE #76: make use of inmanta.module.Module.get_plugin_files for module loading
Currently, `InmantaPluginsImportLoader.get_plugin_files_for_module` duplicates the code in `inmanta.module.Module.get_plugin_files`. It's caller should call the method from core instead.

This method is only introduced in core 2020.4, so this issue is blocked by that release.

### LINKED ISSUE #236: modules v2 - plugin loading compatibility
- [x] implement #76 or reflect the plugin loading changes in core some other way (see discussion on #76)
- [x] Document that both the `inmanta_plugins` fixture and `project.get_plugin_function` support v2 modules but are not required (and not the recommended approach) to import them. v2 modules (that are dependencies of the module being tested) can just be implemented top-level.
- [x] add tests for top-level imports, `inmanta_plugins` fixture and `project.get_plugin_function` for v2 modules

(based on #209)

### PR DESCRIPTION
**Pull request by sanderr**
# Description

Paired with inmanta/inmanta-core#3460.

This pull request makes sure `pytest-inmanta` loads projects in such a way that top level imports remain valid, while also making sure changes to the `unittest` and similar modules are reflected. The tests in the `examples` directory illustrate these goals.

closes #236 
closes #76

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] ~~Changelog entry~~
- [x] Code is clear and sufficiently documented
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: pytest_inmanta/plugin.py [hunk 1]
```diff
@@ -489,26 +464,51 @@ def _create_project_and_load(self, model: str) -> module.Project:
...
             # more recent versions of core require explicit modules installation (ISO5+)
             test_project.install_modules()
         test_project.load()
+        currently_registered_plugins: typing.Mapping[str, typing.Type[plugins.Plugin]] = plugins.PluginMeta.get_functions()
+        loaded_mod_ns_pattern: typing.Pattern[str] = re.compile(
+            "(" + "|".join(re.escape(mod) for mod in test_project.modules.keys()) + ")::"
+        )
+        for fq_plugin_name, plugin in self._registered_plugins.items():
+            if fq_plugin_name not in currently_registered_plugins and loaded_mod_ns_pattern.match(fq_plugin_name):
+                plugins.PluginMeta.add_function(plugin)
```
Comment (wouterdb) [inline pytest_inmanta/plugin.py:510]: I see, this works because `project.load()` builds the ast, but plugins are registered in in 'compiler.compile`

**Review by wouterdb — APPROVED** [review]
Nice

The only comment I have is that modification to  `unittest` virtual module are not reflected anymore

File: pytest_inmanta/plugin.py (hunk 1 above)
Comment (sanderr) [reply pytest_inmanta/plugin.py:510]: Almost. Plugins are registered in `project.load()` as well, but they are added to the namespace only in `compiler.compile`. So by reregistering them after `project.load()`, we mimic the normal behavior.

Comment (sanderr) [reply pytest_inmanta/plugin.py:510]: Almost. Plugins are registered in `project.load()` as well, but they are added to the namespace only in `compiler.compile`. So by reregistering them after `project.load()`, we mimic the normal behavior.

**Review by wouterdb — APPROVED** [review]
very nice!

**Review by arnaudsjs — APPROVED** [review]

**Pull request merged**